File: android/app/src/main/kotlin/ec/cacaotrace/ui/pantallas/DetalleLote.kt
package ec.cacaotrace.ui.pantallas

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AcUnit
import androidx.compose.material.icons.filled.Balance
import androidx.compose.material.icons.filled.Block
import androidx.compose.material.icons.filled.Egg
import androidx.compose.material.icons.filled.GridOn
import androidx.compose.material.icons.filled.History
import androidx.compose.material.icons.filled.LocalShipping
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material.icons.filled.PictureAsPdf
import androidx.compose.material.icons.filled.QrCode2
import androidx.compose.material.icons.filled.Thermostat
import androidx.compose.material.icons.filled.Timeline
import androidx.compose.material.icons.filled.Warehouse
import androidx.compose.material.icons.filled.WbSunny
import androidx.compose.material3.Card
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Tab
import androidx.compose.material3.TabRow
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavHostController
import ec.cacaotrace.ContenedorApp
import ec.cacaotrace.datos.repositorios.HechoLote
import ec.cacaotrace.datos.repositorios.LoteCompleto
import ec.cacaotrace.nucleo.calculo.PasoBalance
import ec.cacaotrace.ui.ColoresEstado
import ec.cacaotrace.ui.Rutas
import ec.cacaotrace.ui.comun.Aviso
import ec.cacaotrace.ui.comun.BarraSuperior
import ec.cacaotrace.ui.comun.BotonGrande
import ec.cacaotrace.ui.comun.EstadoVacio
import ec.cacaotrace.ui.comun.FilaDato
import ec.cacaotrace.ui.comun.Formato
import ec.cacaotrace.ui.comun.TarjetaSeccion
import kotlin.math.abs

/**
 * Detalle del lote: línea de tiempo, balance de masa y lo que toca hacer en la
 * etapa actual (RF-LOT-06, RF-LOT-08).
 */
@Composable
fun PantallaDetalleLote(
    contenedor: ContenedorApp,
    navegacion: NavHostController,
    loteId: String,
) {
    var completo by remember { mutableStateOf<LoteCompleto?>(null) }
    var historia by remember { mutableStateOf<List<HechoLote>>(emptyList()) }
    var balance by remember { mutableStateOf<List<PasoBalance>>(emptyList()) }
    var pestana by remember { mutableIntStateOf(0) }
    var mostrandoQr by remember { mutableStateOf(false) }

    // Se recarga al volver de una pantalla hija: el lote pudo cambiar de etapa.
    val entradaActual = navegacion.currentBackStackEntry
    LaunchedEffect(loteId, entradaActual) {
        completo = contenedor.lotes.completo(loteId)
        historia = contenedor.lotes.lineaDeTiempo(loteId)
        balance = contenedor.lotes.balance(loteId)
    }

    val c = completo ?: return

    if (mostrandoQr) {
        DialogoQr(c.lote.codigo) { mostrandoQr = false }
    }

    Scaffold(
        topBar = {
            BarraSuperior(c.lote.codigo, navegacion) {
                IconButton(onClick = { mostrandoQr = true }) {
                    Icon(Icons.Default.QrCode2, contentDescription = "Código QR")
                }
                IconButton(onClick = { navegacion.navigate(Rutas.reporte(loteId)) }) {
                    Icon(Icons.Default.PictureAsPdf, contentDescription = "Reporte")
                }
            }
        },
    ) { relleno ->
        Column(Modifier.fillMaxSize().padding(relleno)) {
            TabRow(selectedTabIndex = pestana) {
                listOf("Qué sigue", "Historia", "Balance").forEachIndexed { i, titulo ->
                    Tab(
                        selected = pestana == i,
                        onClick = { pestana = i },
                        text = { Text(titulo) },
                    )
                }
            }
            when (pestana) {
                0 -> QueSigue(c, navegacion)
                1 -> Historia(historia)
                else -> Balance(balance)
            }
        }
    }
}

@Composable
private fun QueSigue(c: LoteCompleto, navegacion: NavHostController) {
    val loteId = c.lote.id

    LazyColumn(contentPadding = PaddingValues(16.dp)) {
        item {
            if (c.lote.ventaBloqueada) {
                Aviso(
                    titulo = "Venta bloqueada",
                    texto = c.lote.motivoBloqueo.ifBlank {
                        "Este lote no se puede vender hasta resolverlo."
                    },
                    color = ColoresEstado.problema,
                    icono = Icons.Default.Block,
                )
            }

            TarjetaSeccion(c.lote.estado.etiqueta, icono = Icons.Default.Timeline) {
                Column {
                    Text(c.lote.estado.queSigue, fontSize = 16.sp)
                    Spacer(Modifier.height(12.dp))
                    c.finca?.let { FilaDato("Finca", it.nombre) }
                    FilaDato("Variedad", c.lote.variedad)
                    c.recepcion?.let { FilaDato("Mazorcas", "${it.mazorcasTotal}") }
                    c.apertura?.let { FilaDato("Baba", Formato.kg(it.kgBaba)) }
                    c.secado?.kgSeco?.let { FilaDato("Grano seco", Formato.kg(it)) }
                    c.pruebaCorteFinal?.let { p ->
                        FilaDato(
                            "Prueba de corte",
                            p.resultado,
                            color = if (p.conforme) {
                                ColoresEstado.bien
                            } else {
                                ColoresEstado.atencion
                            },
                        )
                    }
                }
            }

            if (c.alertasAbiertas > 0) {
                Aviso(
                    texto = if (c.alertasAbiertas == 1) {
                        "Hay 1 alerta abierta en este lote"
                    } else {
                        "Hay ${c.alertasAbiertas} alertas abiertas en este lote"
                    },
                    accion = {
                        OutlinedButton(
                            onClick = { navegacion.navigate(Rutas.alertas(loteId)) },
                        ) { Text("Ver") }
                    },
                )
            }

            Spacer(Modifier.height(8.dp))

            if (c.recepcion == null) {
                BotonGrande(
                    texto = "Registrar la recepción",
                    subtitulo = "Mazorcas, peso y días de reposo",
                    icono = Icons.Default.LocalShipping,
                ) { navegacion.navigate(Rutas.recepcion(loteId)) }
            } else {
                BotonGrande(
                    texto = if (c.apertura == null) "Abrir las mazorcas" else "Ver la apertura",
                    subtitulo = c.apertura?.let { Formato.kg(it.kgBaba) + " de baba" }
                        ?: "Registrar kg de baba",
                    icono = Icons.Default.Egg,
                ) { navegacion.navigate(Rutas.apertura(loteId)) }
                Spacer(Modifier.height(12.dp))

                BotonGrande(
                    texto = if (c.fermentacion == null) {
                        "Empezar la fermentación"
                    } else {
                        "Fermentación"
                    },
                    subtitulo = if (c.fermentacion == null) {
                        "Elegir fermentador y masa"
                    } else {
                        "Temperatura, volteos y olor"
                    },
                    icono = Icons.Default.Thermostat,
                ) { navegacion.navigate(Rutas.fermentacion(loteId)) }
                Spacer(Modifier.height(12.dp))

                BotonGrande(
                    texto = if (c.secado == null) "Empezar el secado" else "Secado",
                    subtitulo = c.secado?.kgSeco?.let { Formato.kg(it) + " secos" }
                        ?: "Humedad, remociones y moho",
                    icono = Icons.Default.WbSunny,
                ) { navegacion.navigate(Rutas.secado(loteId)) }
                Spacer(Modifier.height(12.dp))

                BotonGrande(
                    texto = "Prueba de corte",
                    subtitulo = c.pruebaCorteFinal?.resultado
                        ?: "Cortar 100 granos y calificar",
                    icono = Icons.Default.GridOn,
                ) { navegacion.navigate(Rutas.pruebaCorte(loteId)) }
                Spacer(Modifier.height(12.dp))

                BotonGrande(
                    texto = "Almacenamiento",
                    subtitulo = "Sacos, QR e inspecciones",
                    icono = Icons.Default.Warehouse,
                ) { navegacion.navigate(Rutas.almacen(loteId)) }
            }

            Spacer(Modifier.height(24.dp))
            OutlinedButton(
                onClick = { navegacion.navigate(Rutas.alertas(loteId)) },
                modifier = Modifier.fillMaxWidth(),
            ) {
                Icon(Icons.Default.Notifications, contentDescription = null)
                Spacer(Modifier.width(8.dp))
                Text("Alertas y correcciones de este lote")
            }
        }
    }
}

@Composable
private fun Historia(hechos: List<HechoLote>) {
    if (hechos.isEmpty()) {
        EstadoVacio(
            icono = Icons.Default.History,
            titulo = "Todavía no hay historia",
            explicacion = "Aquí se irá anotando todo lo que le pase al lote: lecturas, " +
                "volteos, fotos y alertas.",
        )
        return
    }

    LazyColumn(contentPadding = PaddingValues(16.dp)) {
        itemsIndexed(hechos) { i, h ->
            Row {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Canvas(Modifier.size(14.dp).padding(top = 4.dp)) {
                        drawCircle(
                            color = if (h.esAlerta) ColoresEstado.atencion else Color.Gray,
                        )
                    }
                    if (i < hechos.lastIndex) {
                        Canvas(Modifier.width(2.dp).height(52.dp)) {
                            drawRect(Color.LightGray)
                        }
                    }
                }
                Spacer(Modifier.width(12.dp))
                Column(Modifier.padding(bottom = 16.dp)) {
                    Text(
                        "${Formato.fechaHora(h.fecha)} · ${h.etapa}",
                        fontSize = 13.sp,
                        color = ColoresEstado.neutro,
                    )
                    Text(h.titulo, fontSize = 17.sp)
                    if (h.detalle.isNotBlank()) Text(h.detalle, fontSize = 15.sp)
                }
            }
        }
    }
}

@Composable
private fun Balance(pasos: List<PasoBalance>) {
    if (pasos.isEmpty()) {
        EstadoVacio(
            icono = Icons.Default.Balance,
            titulo = "Sin datos para el balance",
            explicacion = "Registra la recepción para empezar a ver cuánto debería rendir " +
                "el lote en cada etapa.",
        )
        return
    }

    LazyColumn(contentPadding = PaddingValues(16.dp)) {
        item {
            Text(
                "Compara lo que obtuviste con lo que era esperable. Una diferencia grande " +
                    "casi siempre es un error al anotar un peso.",
                fontSize = 15.sp,
            )
            Spacer(Modifier.height(12.dp))
        }
        itemsIndexed(pasos) { _, p ->
            Card(Modifier.fillMaxWidth().padding(vertical = 6.dp)) {
                Column(Modifier.padding(16.dp)) {
                    Text(p.etapa, fontSize = 17.sp)
                    Spacer(Modifier.height(8.dp))
                    FilaDato(
                        "Entró",
                        "${Formato.numero(p.entrada)} ${p.unidadEntrada}",
                    )
                    FilaDato(
                        "Esperado",
                        "${Formato.numero(p.salidaEsperada)} ${p.unidadSalida}",
                    )
                    if (p.registrado) {
                        FilaDato(
                            "Obtenido",
                            "${Formato.numero(p.salidaReal!!)} ${p.unidadSalida}",
                            color = if (abs(p.desvioPct!!) > 20) {
                                ColoresEstado.atencion
                            } else {
                                ColoresEstado.bien
                            },
                        )
                        Text(
                            if (p.desvioPct!! >= 0) {
                                "${Formato.numero(p.desvioPct!!, 0)} % por encima de lo esperado"
                            } else {
                                "${Formato.numero(abs(p.desvioPct!!), 0)} % por debajo de lo " +
                                    "esperado"
                            },
                            fontSize = 14.sp,
                        )
                    } else {
                        FilaDato("Obtenido", "sin registrar", color = ColoresEstado.neutro)
                    }
                }
            }
        }
    }
}

/**
 * Dibuja el QR del lote.
 *
 * Se genera con un generador propio y no con una librería para no añadir una
 * dependencia por un único uso; ver [ec.cacaotrace.ui.comun.GeneradorQr].
 */
@Composable
private fun DialogoQr(codigo: String, alCerrar: () -> Unit) {
    androidx.compose.material3.AlertDialog(
        onDismissRequest = alCerrar,
        title = { Text(codigo) },
        text = {
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Box(Modifier.size(220.dp)) {
                    ec.cacaotrace.ui.comun.CodigoQr(codigo, Modifier.fillMaxSize())
                }
                Spacer(Modifier.height(12.dp))
                Text(
                    "Imprime este código y pégalo en el saco. Al escanearlo se abre la " +
                        "historia completa del lote.",
                    fontSize = 15.sp,
                )
            }
        },
        confirmButton = {
            androidx.compose.material3.TextButton(onClick = alCerrar) { Text("Cerrar") }
        },
    )
}

File: referencia/flutter-app/lib/ui/pantallas/detalle_lote.dart
/// Detalle del lote: línea de tiempo, balance de masa y lo que toca hacer
/// en la etapa actual (RF-LOT-06, RF-LOT-08).
library;

import 'package:flutter/material.dart';
import 'package:qr_flutter/qr_flutter.dart';

import '../../datos/repositorios/repositorio_lotes.dart';
import '../../nucleo/calculo/balance_masa.dart';
import '../../nucleo/modelo/etapas.dart';
import '../../servicios.dart';
import '../comun/widgets.dart';
import '../tema.dart';
import 'alertas.dart';
import 'almacen.dart';
import 'apertura.dart';
import 'fermentacion.dart';
import 'prueba_corte.dart';
import 'recepcion.dart';
import 'reporte.dart';
import 'secado.dart';

class PantallaDetalleLote extends StatefulWidget {
  const PantallaDetalleLote({super.key, required this.loteId});
  final String loteId;

  @override
  State<PantallaDetalleLote> createState() => _PantallaDetalleLoteState();
}

class _PantallaDetalleLoteState extends State<PantallaDetalleLote> {
  Future<LoteCompleto?>? _lote;
  Future<List<HechoLote>>? _historia;
  Future<List<PasoBalance>>? _balance;

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _recargar();
  }

  void _recargar() {
    final s = ProveedorServicios.de(context);
    setState(() {
      _lote = s.lotes.completo(widget.loteId);
      _historia = s.lotes.lineaDeTiempo(widget.loteId);
      _balance = s.lotes.balance(widget.loteId);
    });
  }

  Future<void> _abrir(Widget pantalla) async {
    await Navigator.push(
        context, MaterialPageRoute(builder: (_) => pantalla));
    _recargar();
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<LoteCompleto?>(
      future: _lote,
      builder: (context, snapshot) {
        final c = snapshot.data;
        if (c == null) {
          return const Scaffold(
            body: Center(child: CircularProgressIndicator()),
          );
        }

        return Scaffold(
          appBar: AppBar(
            title: Text(c.lote.codigo),
            actions: [
              IconButton(
                tooltip: 'Código QR',
                icon: const Icon(Icons.qr_code_2),
                onPressed: () => _mostrarQr(context, c.lote.codigo),
              ),
              IconButton(
                tooltip: 'Reporte de trazabilidad',
                icon: const Icon(Icons.picture_as_pdf_outlined),
                onPressed: () =>
                    _abrir(PantallaReporte(loteId: widget.loteId)),
              ),
            ],
          ),
          body: DefaultTabController(
            length: 3,
            child: Column(
              children: [
                const TabBar(tabs: [
                  Tab(text: 'Qué sigue'),
                  Tab(text: 'Historia'),
                  Tab(text: 'Balance'),
                ]),
                Expanded(
                  child: TabBarView(
                    children: [
                      _QueSigue(completo: c, abrir: _abrir, recargar: _recargar),
                      _Historia(futuro: _historia),
                      _Balance(futuro: _balance),
                    ],
                  ),
                ),
              ],
            ),
          ),
        );
      },
    );
  }

  void _mostrarQr(BuildContext context, String codigo) {
    showDialog<void>(
      context: context,
      builder: (contexto) => AlertDialog(
        title: Text(codigo),
        content: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            QrImageView(
              data: codigo,
              size: 220,
              backgroundColor: Colors.white,
            ),
            const SizedBox(height: 16),
            const Text(
              'Imprime este código y pégalo en el saco. Al escanearlo se abre '
              'la historia completa del lote.',
              textAlign: TextAlign.center,
              style: TextStyle(fontSize: 15),
            ),
          ],
        ),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(contexto),
            child: const Text('Cerrar'),
          ),
        ],
      ),
    );
  }
}

class _QueSigue extends StatelessWidget {
  const _QueSigue({
    required this.completo,
    required this.abrir,
    required this.recargar,
  });

  final LoteCompleto completo;
  final Future<void> Function(Widget) abrir;
  final VoidCallback recargar;

  @override
  Widget build(BuildContext context) {
    final lote = completo.lote;
    final loteId = lote.id;

    return ListView(
      padding: const EdgeInsets.all(16),
      children: [
        if (lote.ventaBloqueada)
          Aviso(
            color: ColoresEstado.problema,
            icono: Icons.block,
            titulo: 'Venta bloqueada',
            texto: lote.motivoBloqueo.isEmpty
                ? 'Este lote no se puede vender hasta resolverlo.'
                : lote.motivoBloqueo,
          ),
        _CabeceraEstado(completo: completo),
        const SizedBox(height: 8),
        if (completo.alertasAbiertas > 0)
          Aviso(
            texto: completo.alertasAbiertas == 1
                ? 'Hay 1 alerta abierta en este lote'
                : 'Hay ${completo.alertasAbiertas} alertas abiertas en '
                    'este lote',
            accion: FilledButton(
              onPressed: () => abrir(PantallaAlertas(loteId: loteId)),
              child: const Text('Ver'),
            ),
          ),
        const SizedBox(height: 8),

        // Acciones de la etapa actual, primero la que toca.
        if (completo.recepcion == null)
          BotonGrande(
            texto: 'Registrar la recepción',
            subtitulo: 'Mazorcas, peso y días de reposo',
            icono: Icons.local_shipping_outlined,
            onPressed: () => abrir(PantallaRecepcion(loteId: loteId)),
          )
        else ...[
          BotonGrande(
            texto: completo.apertura == null
                ? 'Abrir las mazorcas'
                : 'Ver la apertura',
            subtitulo: completo.apertura == null
                ? 'Registrar kg de baba'
                : '${completo.apertura!.kgBaba} kg de baba',
            icono: Icons.egg_outlined,
            onPressed: () => abrir(PantallaApertura(loteId: loteId)),
          ),
          const SizedBox(height: 12),
          BotonGrande(
            texto: completo.fermentacion == null
                ? 'Empezar la fermentación'
                : 'Fermentación',
            subtitulo: completo.fermentacion == null
                ? 'Elegir fermentador y masa'
                : 'Temperatura, volteos y olor',
            icono: Icons.thermostat,
            onPressed: () => abrir(PantallaFermentacion(loteId: loteId)),
          ),
          const SizedBox(height: 12),
          BotonGrande(
            texto: completo.secado == null ? 'Empezar el secado' : 'Secado',
            subtitulo: completo.secado?.fin != null
                ? '${completo.secado!.kgSeco ?? 0} kg secos'
                : 'Humedad, remociones y moho',
            icono: Icons.wb_sunny_outlined,
            onPressed: () => abrir(PantallaSecado(loteId: loteId)),
          ),
          const SizedBox(height: 12),
          BotonGrande(
            texto: 'Prueba de corte',
            subtitulo: completo.pruebaCorteFinal?.resultado ??
                'Cortar 100 granos y calificar',
            icono: Icons.grid_on,
            onPressed: () => abrir(PantallaPruebaCorte(loteId: loteId)),
          ),
          const SizedBox(height: 12),
          BotonGrande(
            texto: 'Almacenamiento',
            subtitulo: 'Sacos, QR e inspecciones',
            icono: Icons.warehouse_outlined,
            onPressed: () => abrir(PantallaAlmacen(loteId: loteId)),
          ),
        ],
        const SizedBox(height: 24),
        OutlinedButton.icon(
          icon: const Icon(Icons.notifications_outlined),
          label: const Text('Alertas y correcciones de este lote'),
          onPressed: () => abrir(PantallaAlertas(loteId: loteId)),
        ),
      ],
    );
  }
}

class _CabeceraEstado extends StatelessWidget {
  const _CabeceraEstado({required this.completo});
  final LoteCompleto completo;

  @override
  Widget build(BuildContext context) {
    final c = completo;
    final prueba = c.pruebaCorteFinal;

    return TarjetaSeccion(
      titulo: c.lote.estado.etiqueta,
      icono: Icons.timeline,
      hijos: [
        Text(c.lote.estado.queSigue, style: const TextStyle(fontSize: 16)),
        const Divider(height: 24),
        if (c.finca != null) FilaDato('Finca', c.finca!.nombre),
        FilaDato('Variedad', c.lote.variedad),
        if (c.recepcion != null)
          FilaDato('Mazorcas', '${c.recepcion!.mazorcasTotal}'),
        if (c.apertura != null)
          FilaDato('Baba', '${c.apertura!.kgBaba} kg'),
        if (c.secado?.kgSeco != null)
          FilaDato('Grano seco', '${c.secado!.kgSeco} kg'),
        if (prueba != null)
          FilaDato(
            'Prueba de corte',
            prueba.resultado,
            color: prueba.conforme
                ? ColoresEstado.bien
                : ColoresEstado.atencion,
          ),
      ],
    );
  }
}

class _Historia extends StatelessWidget {
  const _Historia({required this.futuro});
  final Future<List<HechoLote>>? futuro;

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<List<HechoLote>>(
      future: futuro,
      builder: (context, snapshot) {
        final hechos = snapshot.data ?? const [];
        if (hechos.isEmpty) {
          return const EstadoVacio(
            icono: Icons.history,
            titulo: 'Todavía no hay historia',
            explicacion: 'Aquí se irá anotando todo lo que le pase al lote: '
                'lecturas, volteos, fotos y alertas.',
          );
        }
        return ListView.builder(
          padding: const EdgeInsets.all(16),
          itemCount: hechos.length,
          itemBuilder: (context, i) {
            final h = hechos[i];
            return Row(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Column(
                  children: [
                    Container(
                      width: 14,
                      height: 14,
                      margin: const EdgeInsets.only(top: 6),
                      decoration: BoxDecoration(
                        shape: BoxShape.circle,
                        color: h.esAlerta
                            ? ColoresEstado.atencion
                            : Theme.of(context).colorScheme.primary,
                      ),
                    ),
                    if (i < hechos.length - 1)
                      Container(
                        width: 2,
                        height: 52,
                        color: Theme.of(context).colorScheme.outlineVariant,
                      ),
                  ],
                ),
                const SizedBox(width: 12),
                Expanded(
                  child: Padding(
                    padding: const EdgeInsets.only(bottom: 16),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          '${_fechaHora(h.fecha)} · ${h.etapa}',
                          style: TextStyle(
                            fontSize: 13,
                            color: Theme.of(context)
                                .colorScheme
                                .onSurfaceVariant,
                          ),
                        ),
                        Text(h.titulo,
                            style: const TextStyle(
                                fontSize: 17, fontWeight: FontWeight.w600)),
                        if (h.detalle.isNotEmpty)
                          Text(h.detalle,
                              style: const TextStyle(fontSize: 15)),
                      ],
                    ),
                  ),
                ),
              ],
            );
          },
        );
      },
    );
  }
}

class _Balance extends StatelessWidget {
  const _Balance({required this.futuro});
  final Future<List<PasoBalance>>? futuro;

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<List<PasoBalance>>(
      future: futuro,
      builder: (context, snapshot) {
        final pasos = snapshot.data ?? const [];
        if (pasos.isEmpty) {
          return const EstadoVacio(
            icono: Icons.balance,
            titulo: 'Sin datos para el balance',
            explicacion: 'Registra la recepción para empezar a ver cuánto '
                'debería rendir el lote en cada etapa.',
          );
        }
        return ListView(
          padding: const EdgeInsets.all(16),
          children: [
            const Text(
              'Compara lo que obtuviste con lo que era esperable. Una '
              'diferencia grande casi siempre es un error al anotar un peso.',
              style: TextStyle(fontSize: 15),
            ),
            const SizedBox(height: 16),
            for (final p in pasos)
              Card(
                child: Padding(
                  padding: const EdgeInsets.all(16),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(p.etapa,
                          style: const TextStyle(
                              fontSize: 17, fontWeight: FontWeight.w600)),
                      const SizedBox(height: 8),
                      FilaDato('Entró',
                          '${p.entrada.toStringAsFixed(1)} ${p.unidadEntrada}'),
                      FilaDato(
                        'Esperado',
                        '${p.salidaEsperada.toStringAsFixed(1)} '
                            '${p.unidadSalida}',
                      ),
                      if (p.registrado)
                        FilaDato(
                          'Obtenido',
                          '${p.salidaReal!.toStringAsFixed(1)} '
                              '${p.unidadSalida}',
                          color: p.desvioPct!.abs() > 20
                              ? ColoresEstado.atencion
                              : ColoresEstado.bien,
                        )
                      else
                        const FilaDato('Obtenido', 'sin registrar',
                            color: ColoresEstado.neutro),
                      if (p.registrado)
                        Text(
                          p.desvioPct! >= 0
                              ? '${p.desvioPct!.toStringAsFixed(0)} % por '
                                  'encima de lo esperado'
                              : '${p.desvioPct!.abs().toStringAsFixed(0)} % '
                                  'por debajo de lo esperado',
                          style: const TextStyle(fontSize: 14),
                        ),
                    ],
                  ),
                ),
              ),
          ],
        );
      },
    );
  }
}

String _fechaHora(DateTime f) =>
    '${f.day.toString().padLeft(2, '0')}/'
    '${f.month.toString().padLeft(2, '0')} '
    '${f.hour.toString().padLeft(2, '0')}:'
    '${f.minute.toString().padLeft(2, '0')}';
